### PR TITLE
[Enhancement] Reintroduce the `MapStore` interface and `SimpleMap` and make the `Badger` a submodules

### PR DESCRIPTION
**Pull request by h5law**
### Human Summary

This PR reintroduces the `MapStore` interface and `SimpleMap` implementation of a basic in-memory key-value store and makes the badger `BadgerStore` interface and tests a submodule, in the case where people do not want this dependency.

It also updates the Makefile targets and workflows to account for this change in project structure.

### Summary generated by Reviewpad on 03 Jan 24 22:49 UTC
This pull request includes several changes across multiple files:

1. Modifications to the `bulk_test.go` file:
   - Importing a new package `simplemap`.
   - Modifying the creation of variables `smn` and `smv` using `simplemap.NewSimpleMap()` function instead of `NewKVStore`.
   - Updating error handling in the `bulkOperations` function.

2. Addition of a new file `interfaces.go` in the `kvstore` package:
   - Defines the `MapStore` interface for key-value store functionalities.

3. Addition of a new test file `smt_example_test.go` demonstrating the usage of Sparse Merkle Trie (SMT).
   
4. Changes to the `fuzz_test.go` file:
   - Adding an import statement for `github.com/pokt-network/smt/kvstore/simplemap`.
   - Fixing typos in comments and error messages.
   - Updating error comparisons and messages.
   - Relocating the `smn.Stop()` statement.

5. Modifications to the README.md file:
   - Addition of a new section about testing the 'badger' submodule.

6. Changes to the `smt.md` file:
   - Addition of subsections under the `Database` section.
   - Updates to descriptions and explanations for the `MapStore` and `KVStore` interfaces.
   - Addition of details about the `SimpleMap` and `Badger` submodules.
   - Removal of the `Example` section.

7. Addition of the `mapstore.md` file:
   - Introduction to the `MapStore` interface and its implementations (`SimpleMap` and `BadgerV4`).

8. Renaming and restructuring of packages and test names in various files.

9. Renaming and updating the `kvstore.md` file to `badger-store.md`.

10. Changes to `smt_proofs_test.go`:
    - Imports and type modifications related to the `kvstore` and `simplemap` packages.

11. Changes to `proofs.go`:
    - Removal of the `init()` function.
    - Declaration of the `SparseMerkleProof` struct.

12. Addition of the `godoc.go` file to the `kvstore` directory.

13. Changes to the Makefile:
    - Modifications to targets, including removal, addition, and updates to descriptions.

14. Addition of a new test case in the `smst_example_test.go` file demonstrating the usage of the `simplemap` package.

15. Addition of a new file `simplemap.go` implementing the `simplemap` package.

16. Changes to the `smst_utils_test.go` file:
    - Modifications to import statements and variable types.
    - Improvements in comments and function explanations.

17. Changes to the `godoc.go` file in the `smt` package:
    - Updates to package description and optimizations mentioned.

18. Changes to the `smst_proofs_test.go` file:
    - Modifications to imports, variable types, and function calls.

19. Changes to the `godoc.go` file:
    - Updates to package documentation, optimizations, and features mentioned.

20. Modifications to the `godoc.go` file:
    - Addition of import statements and replacements of variable types.

Please review these changes in detail and provide any necessary feedback.

## Issue

Fixes N/A

- #28 raised this as an issue

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [x] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make test_all`
- [ ] **Run all/relevant benchmarks (if optimising)**: `make benchmark_{all | suite name}`

## Required Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code ([`godoc` format comments](https://go.dev/blog/godoc) see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))

### If Applicable Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated any relevant README(s)/documentation and left TODOs throughout the codebase
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

### CODE REVIEW

**Comment by h5law** [comment]
@muXxer Let me know if this is what you were looking for from your PR

File: docs/smt.md
Comment (Olshansk) [inline docs/smt.md:453]:
```suggestion
By default, this library provides a simple interface (`MapStore`) and a simple
in-memory key-value database found in [`simplemap.go`](../kvstore/simplemap.go).
However, any key-value store implementing this interface can be used as the
node store to back the trie.
```

File: docs/smt.md
Comment (Olshansk) [inline docs/smt.md:464]:
```suggestion
In addition to providing the `MapStore` and `SimpleMap` interfaces and
implementations, the `smt` library also provides wrappers around other
key-value databases as submodules with more fully-featured interfaces
that can be used outside of backing key-value engines for tries. These
submodules can be found in the [`kvstore`](../kvstore/) directory.
```

File: kvstore/simplemap.go
Comment (Olshansk) [inline kvstore/simplemap.go:1]: For consistency with badger (and other key-value stores we'll add in the future), I think we should create a package called `simplemap` and have an associated `interface.go` file.

It's a NIT but trying to set a pattern for others to follow.

File: kvstore/simplemap.go
Comment (Olshansk) [inline kvstore/simplemap.go:6]:
```suggestion
// MapStore defines an interface that represents a key-value store that backs
// the SM(S)T. It is the minimum viable subset of functionality a key-value
// store requires in order to back an SM(S)T.
```

File: kvstore/simplemap.go
Comment (Olshansk) [inline kvstore/simplemap.go:18]: 1. Consider splitting the implementation & interface into separate files (see comment above)
2. Assert struct implements the interface (set patterns others will copy-paste)

File: kvstore/godoc.go
Comment (Olshansk) [inline kvstore/godoc.go:2]: Left another comment, but we'll need to update the godoc if we make a submodule for simplemap.

File: godoc.go
Comment (Olshansk) [inline godoc.go:5]:
```
// Package smt provides an implementation of a Sparse Merkle Trie for a
// key-value map.
// 
// The trie implements the same optimizations specified in the JMT
// whitepaper to account for empty and single-node subtrees. Unlike the
// JMT, it only supports binary trees and does not optimise for RockDB
// on-disk storage.
// 
// This package implements novel features that include native in-node
// weight sums, as well as support for ClosestProof mechanics.
```

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: Makefile
Comment (h5law) [reply Makefile:43]: This doesn't work as kvstore contains a submodule

File: .github/workflows/test.yml
Comment (h5law) [reply .github/workflows/test.yml:48]: There is a non benchmark test in the benchmarks dir and this is the easiest way to exclude it - without a subprocess running go list | grep ...

File: kvstore/badger/interface.go
Comment (h5law) [reply kvstore/badger/interface.go:14]: This interface is specific to the badger DB and isn't general purpose. That being said it bad be used as the node store of the SMT as it implements `MapStore`

File: kvstore/simplemap.go
Comment (h5law) [reply kvstore/simplemap.go:1]: I think badger is the pattern and simplemap is never likely to be used in prod. I am against making MapStore a submodule it needs to be accessable to the tries

File: smt_test.go
Comment (h5law) [reply smt_test.go:15]: preimage === value inserted after hashing

value === value we interted

preimage === value IFF non value hahsing tree

digest is a hash simple but preimage and value are different

Comment (h5law) [reply smt_test.go:15]: This is a simplification ofc

File: kvstore/simplemap.go
Comment (Olshansk) [reply kvstore/simplemap.go:1]: Going to push back against this. The way I see it is:

- `Simplemap` is an example (and the default) that won't be used in prod
- `Badger` is the first `prod` implementation
- There should be a share (base) interface

Happy to jump on a call and discuss more if you think it would help.

File: kvstore/badger/interface.go
Comment (Olshansk) [reply kvstore/badger/interface.go:14]: > This interface is specific to the badger DB and isn't general purpose

In that case, we should probably update the name & description to reflect that.

For example `BadgerKVStore` with a comment that says `A superset of the default minimal KVstore ...`.

Happy to jump on a call to discuss if you think it'll require a longer back-and-forth.

File: smt_test.go
Comment (Olshansk) [reply smt_test.go:15]: #PUC or #PUD (Please Update Documentation)

**Comment by h5law** [comment]
@Olshansk what do you think of the new structure re: simplemap making it a submodule and separating the `MapStore` interface from any implementations

File: Makefile
Comment (Olshansk) [inline Makefile:54]:
```suggestion
###############
### go helpers ###
###############
.PHONY: go_mod_tidy
go_mod_tidy:  ## runs go mod tidy for all (sub)modules
```

Comment (Olshansk) [reply Makefile:54]: It'll extend to other go helpers

File: smt_example_test.go
Comment (Olshansk) [inline smt_example_test.go:1]: Should this not be package `smt`?

File: smst_example_test.go
Comment (Olshansk) [inline smst_example_test.go:1]:
```suggestion
package smt
```

**Review by Olshansk — CHANGES_REQUESTED** [review]
A couple small nits and outstanding comments from the previous review, but should be g2g after

File: smst_example_test.go
Comment (h5law) [reply smst_example_test.go:1]: No it needs to have the _test to act as an external application

File: smt_example_test.go
Comment (h5law) [reply smt_example_test.go:1]: See https://github.com/pokt-network/smt/pull/33#discussion_r1440698387

File: kvstore/badger/interface.go
Comment (h5law) [reply kvstore/badger/interface.go:14]: No this makes sense done

File: Makefile
Comment (Olshansk) [inline Makefile:48]: Do we have to check for godoc here?

File: benchmarks/bench_smt_test.go
Comment (Olshansk) [inline benchmarks/bench_smt_test.go:237]: Make sure to s/name/desc here and below

File: benchmarks/proof_sizes_test.go
Comment (Olshansk) [inline benchmarks/proof_sizes_test.go:1]:
```suggestion
//go:build benchmark
```

File: Makefile
Comment (Olshansk) [inline Makefile:59]: Make sure to add the benchmark tag to these helpers

File: benchmarks/bench_smt_test.go
Comment (Olshansk) [inline benchmarks/bench_smt_test.go:237]: s/name/desc here and below

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: Makefile
Comment (h5law) [reply Makefile:48]: meant to check go version lmao

File: Makefile
Comment (Olshansk) [inline Makefile:22]:
```suggestion
# Internal helper target - check if godoc is installed
```

File: Makefile
Comment (Olshansk) [inline Makefile:28]: Was this intentional?

**Review by Olshansk — APPROVED** [review]
Left two more comments, but approving so we can get this in.

**Comment by Olshansk** [comment]
@h5law Make sure to resolve conflicts with main and fix the indents.

File: Makefile
Comment (h5law) [reply Makefile:28]: Yeah this level of intentation is required my Makefile syntax its anoying

File: docs/smt.md
Comment (Olshansk) [inline docs/smt.md:475]:
```suggestion
Although it is a submodule, it is ideal for simple, testing or non-production
```

File: docs/kvstore.md
Comment (Olshansk) [inline docs/kvstore.md:1]: rename the file to `mapstore.md` from `kvstore.md`

File: docs/kvstore.md
Comment (Olshansk) [inline docs/kvstore.md:14]:
```suggestion
retrieve key-value pairs. It is intentionally simple and minimalistic so as to 
enable different key-value engines to implement and back the trie database.
```

File: docs/kvstore.md
Comment (Olshansk) [inline docs/kvstore.md:26]:
```suggestion
`simplemap` is a simple kv-store shipped with the SM(S)T. The SMT library
can be used without it as long as the selected node store adheres to the 
`Mapstore` interface.
```

File: docs/kvstore.md
Comment (Olshansk) [inline docs/kvstore.md:28]:
```suggestion
This library is recommended for development, testing and exploration purposes. 
This library SHOULD NOT be used in production.
```

File: docs/kvstore.md
Comment (Olshansk) [inline docs/kvstore.md:36]:
```suggestion
This library provides a wrapper around [dgraph-io/badger](https://github.com/dgraph-io/badger)
to adhere to the `MapStore` interface. See the [full documentation](./badger-store.md) for additional
functionality and implementation details.
```

**Review by Olshansk — APPROVED** [review]
Made some doc improvements. Please review before merging.

**Pull request merged**
